sql/core/src/main/scala/org/apache/spark/sql/execution/AggregatingAccumulator.scala

```diff
@@ -205,6 +205,10 @@ class AggregatingAccumulator private(
     buffer = other.buffer
     joinedRow = other.joinedRow
   }
+
+  private[execution] def addState(v: InternalRow): Unit = {
```

Comment:
how about
```
def addState(s: AggregatingAccumulator): Unit = {
  if (buffer == null) {
    buffer = s.buffer
    joinedRow = s.joinedRow
  } else {
    add(v.value)
  }
}
```

Reply:
note: we should call `merge` not `add`, as the state to add is another computed aggregator, not an input row.

Reply:
In fact, it's my first idea of `addState` but didn't work. I investigated the reason and found that `merge` is not intended to be called from Executor.

https://github.com/apache/spark/blob/33f26275f4d65f54e68f38ba0d795396a5a4d2f4/sql/core/src/main/scala/org/apache/spark/sql/execution/AggregatingAccumulator.scala#L96-L108

Do you think we can relax the rule here?

Reply:
I see, so we need a per-task `AggregatingAccumulator` instance and keeping calling `add` during the task lifetime.

The problem is that we need a temp `AggregatingAccumulator` instance to delay the accumulator updates until the end of the task, and it's a bit hard to keep this `AggregatingAccumulator` per task.

Let me think more about this and see if I can come up with a better idea.

Reply:
See https://github.com/apache/spark/pull/32786/files#r646703026

Reply:
See https://github.com/apache/spark/pull/32786/files#r646710820